Refactor the archiving process - WIP

ke4 opened the pull request:
Changes:
- Refactor archiving to accommodate ENA submissions.
- ENA submissions has some differences from other archive submissions.
  - need to deal with multiple entity type.
  - only 1 REST call even for multiply entity types (resource types)
  - if we submit more than 1 from an entity type they should go into the same XML payload file into a SET
- I also did some code cleaning, but probably we have to do more cleaning later.
- Unit test also has been refactored as code changed 

I tested the submission process with the same dataset we are using for the integration tests (SS2 dataset).
I got the above result:

```
2022-01-28 18:43:09,208 - __main__ - INFO - Archiving has started in local environment for submission: 6f8de17d-ee87-4968-9373-0936d0358e8a. This could take some time. Please, be patience.
2022-01-28 18:43:16,785 - __main__ - INFO - Response: {'biosamples': {'CREATED': [{'data': {'_links': {'curationDomain': {'href': 'https://wwwdev.ebi.ac.uk/biosamples/samples{?curationdomain}', 'templated': True}, 'curationLink': {'href': 'https://wwwdev.ebi.ac.uk/biosamples/samples/SAMEA8832279/curationlinks/{hash}', 'templated': True}, 'curationLinks': {'href': 'https://wwwdev.ebi.ac.uk/biosamples/samples/SAMEA8832279/curationlinks'}, 'self': {'href': 'https://wwwdev.ebi.ac.uk/biosamples/samples'}, 'structuredData': {'href': 'https://wwwdev.ebi.ac.uk/biosamples/structureddata/SAMEA8832279'}}, 'accession': 'SAMEA8832279', 'characteristics': {'Biomaterial Core - Biomaterial Id': [{'text': 'cell_ID_1'}], 'HCA Biomaterial Type': [{'text': 'cell_suspension'}], 'HCA Biomaterial UUID': [{'text': '5f39d18e-cfe1-4ea1-bd4d-6ba92b0bc411'}], 'organism': [{'ontologyTerms': ['http://purl.obolibrary.org/obo/NCBITaxon_9606'], 'text': 'Homo sapiens'}], 'project': [{'text': 'Human Cell Atlas'}]}, 'create': '2022-01-28T18:43:11.297Z', 'domain': 'self.hca-test', 'name': 'This is a dummy cell', 'release': '2022-01-25T12:09:27Z', 'releaseDate': '2022-01-25', 'submitted': '2022-01-28T18:43:11.297Z', 'submittedDate': '2022-01-28', 'submittedVia': 'JSON_API', 'taxId': 9606, 'update': '2022-01-28T18:43:11.297Z', 'updateDate': '2022-01-28', 'uuid': '5f39d18e-cfe1-4ea1-bd4d-6ba92b0bc411'}, 'entity_type': 'biomaterials', 'error_messages': None, 'is_update': False}, {'data': {'_links': {'curationDomain': {'href': 'https://wwwdev.ebi.ac.uk/biosamples/samples{?curationdomain}', 'templated': True}, 'curationLink': {'href': 'https://wwwdev.ebi.ac.uk/biosamples/samples/SAMEA8832280/curationlinks/{hash}', 'templated': True}, 'curationLinks': {'href': 'https://wwwdev.ebi.ac.uk/biosamples/samples/SAMEA8832280/curationlinks'}, 'self': {'href': 'https://wwwdev.ebi.ac.uk/biosamples/samples'}, 'structuredData': {'href': 'https://wwwdev.ebi.ac.uk/biosamples/structureddata/SAMEA8832280'}}, 'accession': 'SAMEA8832280', 'characteristics': {'Biomaterial Core - Biomaterial Id': [{'text': 'specimen_ID_1'}], 'HCA Biomaterial Type': [{'text': 'specimen_from_organism'}], 'HCA Biomaterial UUID': [{'text': '766da0f0-0d14-4961-9196-e9c0062deee2'}], 'organism': [{'ontologyTerms': ['http://purl.obolibrary.org/obo/NCBITaxon_9606'], 'text': 'Homo sapiens'}], 'project': [{'text': 'Human Cell Atlas'}]}, 'create': '2022-01-28T18:43:11.472Z', 'domain': 'self.hca-test', 'name': 'This is a dummy specimen', 'release': '2022-01-25T12:09:27Z', 'releaseDate': '2022-01-25', 'submitted': '2022-01-28T18:43:11.472Z', 'submittedDate': '2022-01-28', 'submittedVia': 'JSON_API', 'taxId': 9606, 'update': '2022-01-28T18:43:11.472Z', 'updateDate': '2022-01-28', 'uuid': '766da0f0-0d14-4961-9196-e9c0062deee2'}, 'entity_type': 'biomaterials', 'error_messages': None, 'is_update': False}, {'data': {'_links': {'curationDomain': {'href': 'https://wwwdev.ebi.ac.uk/biosamples/samples{?curationdomain}', 'templated': True}, 'curationLink': {'href': 'https://wwwdev.ebi.ac.uk/biosamples/samples/SAMEA8832281/curationlinks/{hash}', 'templated': True}, 'curationLinks': {'href': 'https://wwwdev.ebi.ac.uk/biosamples/samples/SAMEA8832281/curationlinks'}, 'self': {'href': 'https://wwwdev.ebi.ac.uk/biosamples/samples'}, 'structuredData': {'href': 'https://wwwdev.ebi.ac.uk/biosamples/structureddata/SAMEA8832281'}}, 'accession': 'SAMEA8832281', 'characteristics': {'Biomaterial Core - Biomaterial Id': [{'text': 'donor_ID_1'}], 'HCA Biomaterial Type': [{'text': 'donor_organism'}], 'HCA Biomaterial UUID': [{'text': '68e2b9d0-759c-45cc-8d63-0eced61a55e9'}], 'Is Living': [{'text': 'unknown'}], 'Sex': [{'text': 'unknown'}], 'organism': [{'ontologyTerms': ['http://purl.obolibrary.org/obo/NCBITaxon_9606'], 'text': 'Homo sapiens'}], 'project': [{'text': 'Human Cell Atlas'}]}, 'create': '2022-01-28T18:43:11.654Z', 'domain': 'self.hca-test', 'name': 'This is a dummy donor', 'release': '2022-01-25T12:09:27Z', 'releaseDate': '2022-01-25', 'submitted': '2022-01-28T18:43:11.654Z', 'submittedDate': '2022-01-28', 'submittedVia': 'JSON_API', 'taxId': 9606, 'update': '2022-01-28T18:43:11.654Z', 'updateDate': '2022-01-28', 'uuid': '68e2b9d0-759c-45cc-8d63-0eced61a55e9'}, 'entity_type': 'biomaterials', 'error_messages': None, 'is_update': False}]}, 'biostudies': {'CREATED': [{'data': {'accession': 'S-BSST388', 'uuid': 'ad14a919-d7dc-410e-a6eb-31543c682ff5'}, 'entity_type': 'projects', 'error_messages': None, 'is_update': False}]}, 'ena': {'CREATED': [{'data': {'accession': 'ERP135067', 'uuid': 'ad14a919-d7dc-410e-a6eb-31543c682ff5'}, 'entity_type': 'projects', 'error_messages': None, 'is_update': False}, {'data': {'accession': 'ERS10222314', 'uuid': '5f39d18e-cfe1-4ea1-bd4d-6ba92b0bc411'}, 'entity_type': 'biomaterials', 'error_messages': None, 'is_update': False}, {'data': {'accession': 'ERS10222315', 'uuid': '766da0f0-0d14-4961-9196-e9c0062deee2'}, 'entity_type': 'biomaterials', 'error_messages': None, 'is_update': False}, {'data': {'accession': 'ERS10222316', 'uuid': '68e2b9d0-759c-45cc-8d63-0eced61a55e9'}, 'entity_type': 'biomaterials', 'error_messages': None, 'is_update': False}]}}
2022-01-28 18:43:16,785 - __main__ - INFO - You can check the result of archiving in the following webpages:
2022-01-28 18:43:16,787 - __main__ - INFO - {
    "biosamples": {
        "CREATED entities in biosamples": [
            {
                "uuid": "5f39d18e-cfe1-4ea1-bd4d-6ba92b0bc411",
                "url": "https://wwwdev.ebi.ac.uk/biosamples/samples/SAMEA8832279"
            },
            {
                "uuid": "766da0f0-0d14-4961-9196-e9c0062deee2",
                "url": "https://wwwdev.ebi.ac.uk/biosamples/samples/SAMEA8832280"
            },
            {
                "uuid": "68e2b9d0-759c-45cc-8d63-0eced61a55e9",
                "url": "https://wwwdev.ebi.ac.uk/biosamples/samples/SAMEA8832281"
            }
        ]
    },
    "biostudies": {
        "CREATED entities in biostudies": [
            {
                "uuid": "ad14a919-d7dc-410e-a6eb-31543c682ff5",
                "url": "https://wwwdev.ebi.ac.uk/biostudies/studies/S-BSST388"
            }
        ]
    },
    "ena": {
        "CREATED entities in ena": [
            {
                "uuid": "ad14a919-d7dc-410e-a6eb-31543c682ff5",
                "url": "https://wwwdev.ebi.ac.uk/ena/submit/webin/report/studies/ERP135067"
            },
            {
                "uuid": "5f39d18e-cfe1-4ea1-bd4d-6ba92b0bc411",
                "url": "https://wwwdev.ebi.ac.uk/ena/submit/webin/report/samples/ERS10222314"
            },
            {
                "uuid": "766da0f0-0d14-4961-9196-e9c0062deee2",
                "url": "https://wwwdev.ebi.ac.uk/ena/submit/webin/report/samples/ERS10222315"
            },
            {
                "uuid": "68e2b9d0-759c-45cc-8d63-0eced61a55e9",
                "url": "https://wwwdev.ebi.ac.uk/ena/submit/webin/report/samples/ERS10222316"
            }
        ]
    }
}
```

amnonkhen commented on archiver_app.py:
```diff
@@ -84,15 +86,35 @@ def archive():
+            archives_response = direct_archiver.archive_submission(submission_uuid)
             logger.info('Archiving finished with UUID=%s', submission_uuid)
-        except RestErrorException as rest_error:
+        except Exception as rest_error:
```
This error handling section is very long.
Have you considered a function for each case?
Have you considered using an `except` clause for each exception type?
Have you considered using a dataclass or namedtumple for the error response object?

ke4 replied:
Cleaned up

amnonkhen commented on converter/ena/base.py:
```diff
@@ -50,7 +52,9 @@ def _get_entity_content(entity: dict) -> dict:
     def _get_value_by_key_path(entity: dict, key_path: list) -> str:
```
This function could be in a separate module. It is not specific to ena.

ke4 replied:
Moved all of these kind of utility methods for conversion to package level

amnonkhen commented on submit_to_archives.py:
```diff
@@ -71,18 +82,37 @@ def get_header():
+    response_text = response_json
+
+    output = {}
+    for archive_name, archived_items in response_text.items():
```
There a re many nesting levels and conditionals here. This would make it difficult to understand, test and debug.

ke4 replied:
Refactored

amnonkhen commented on submitter/base.py:
```diff
@@ -2,81 +2,136 @@
-        return response, accessions
+        hca_entity_types = ARCHIVE_TO_HCA_ENTITY_MAP[archive_type]
+        converted_entities = []
+        for hca_entity_type in hca_entity_types:
```
See comment about multiple nesting levels and conditionals.

ke4 replied:
Done

amnonkhen commented on submitter/base.py:
```diff
@@ -2,81 +2,136 @@
+        converted_entities = []
+        for hca_entity_type in hca_entity_types:
+            is_update = False
+            if archive_type == 'ENA':
```
When I see several occurrences of `if archive_type == 'ENA'` scattered in the function, it is a hint that the "strategy pattern" could (and probably should) have been used here.

amnonkhen commented on tests/unit/submitter/test_ena_submitter.py:
```diff
@@ -0,0 +1,191 @@
+            return file.read()
+
+
+    # def test_submit_project_converts_and_sends_project(self):
```
IS this commented code on the way out, or about to be fixed?

ke4 replied:
Accidentally forgot to delete it. It is done now. Tx

aaclan-ebi commented on tests/unit/submitter/test_biostudies_submitter.py:
```diff
@@ -1,85 +1,161 @@
+    #     self.submitter._submit_to_archive = MagicMock(return_value={'accession': self.biostudies_accession})
+    #     test_case = self.submission.map_ingest_entity(make_ingest_entity(self.entity_type, random_id(), random_uuid()))
+    #     # When
+    #     result, accession = self.submitter.send_entity(test_case, self.archive_type, self.error_key)
+    #     # Then
+    #     self.assertEqual(CREATED_ENTITY, result)
+    #
+    # def test_submit_project_return_error(self):
+    #     # Given
+    #     error_msg = f'TestCaseError{random_id(3)}'
+    #     self.submitter._submit_to_archive = MagicMock(side_effect=Exception(error_msg))
+    #     test_case = self.submission.map_ingest_entity(
+    #         make_ingest_entity(self.entity_type, random_id(), random_uuid()))
+    #     # When
+    #     result, accession = self.submitter.send_entity(test_case, self.archive_type, self.error_key)
+    #     # Then
+    #     self.assertEqual(ERRORED_ENTITY, result)
+    #     self.assertDictEqual(test_case.get_errors(), {self.error_key: [f'BioStudies Error: {error_msg}']})
+    #
+    # def test_send_all_projects_sorts_projects_into_results(self):
+    #     # Given
+    #     self.submitter.send_entity = MagicMock(side_effect=[
+    #         [UPDATED_ENTITY, "1234"], [CREATED_ENTITY, "5678"], [ERRORED_ENTITY, "9012"]
+    #     ])
+    #     entities = self.map_random_entities(self.submission, self.entity_type, 3)
+    #     calls = []
+    #     call_count = 0
+    #     for entity in entities:
+    #         call_count += 1
+    #         calls.append(call(entity, self.archive_type, self.error_key, {}))
+    #
+    #     # When
+    #     result, accessions = self.submitter.send_all_projects(self.submission)
+    #
+    #     # Then
+    #     self.submitter.send_entity.assert_has_calls(calls, any_order=True)
+    #     self.assertEqual(call_count, self.submitter.send_entity.call_count)
+    #     self.assertEqual(1, len(result[UPDATED_ENTITY]))
+    #     self.assertEqual(1, len(result[CREATED_ENTITY]))
+    #     self.assertEqual(1, len(result[ERRORED_ENTITY]))
```
why are these commented out?

ke4 replied:
Accidentally forgot to delete it. It is done now. Tx

aaclan-ebi reviewed (commented):
Hi @ke4, I believe you can still further extract long bits to smaller functions for readability.

ke4 commented on converter/ena/base.py:
```diff
@@ -50,7 +52,9 @@ def _get_entity_content(entity: dict) -> dict:
         value = entity
         for key in key_path:
-            value = value[key]
+            value = value.get(key, None)
```
Done

ke4 commented on submit_to_archives.py:
```diff
@@ -71,18 +82,37 @@ def get_header():
+    response_text = response_json
+
+    output = {}
+    for archive_name, archived_items in response_text.items():
+        archive_result = {}
+        for status, results in archived_items.items():
+            data = []
+            for result_data in results:
+                base_url = ARCHIVE_URLS[archive_name]
```
Done

ke4 commented on submitter/base.py:
```diff
@@ -2,81 +2,136 @@
+            if archive_type == 'ENA':
+                converted_ena_xml = self.converter.convert_entity_to_xml_str()
+                if len(converted_ena_xml) < 1:
+                    break
```
Done

ke4 commented on submitter/ena_study.py:
```diff
@@ -1,41 +1,88 @@
+                    processed_response.append(
+                        ArchiveResponse(entity_type=entity_type, data=data, is_update=is_update)
+                    )
+
```
Done

ke4 commented on archiver_app.py:
```diff
@@ -84,15 +86,35 @@ def archive():
+            archives_response = {
+                'message': 'Archiving failed.'
             }
+            if isinstance(rest_error, ArchiveException):
```
Cleaned up

aaclan-ebi reviewed (approved):
I think you can still refactor this further using strategy pattern for each archive submitter (like what Amnon suggested) but I think it's a great improvement and since you added tests, you can refactor further later. I'll approve it.

aaclan-ebi commented:
Pls fix the test build

amnonkhen commented:
Thanks, @ke4 for addressing the comments. It looks much better now.
I feel that using the Strategy pattern would make the code much more maintainable and simpler to understand. I urge you to do it, if not in this iteration, in the next one.

amnonkhen commented:
Is there a test that checks that all the ENA files for a certain project are created correctly?

ke4 commented:
> Is there a test that checks that all the ENA files for a certain project are created correctly?

@amnonkhen What kind of ENA files? We are not dealing here with files. Are you talking about sample conversion? Sorry it is not really clear what would you like to know.

ke4 commented:
> Thanks, @ke4 for addressing the comments. It looks much better now. I feel that using the Strategy pattern would make the code much more maintainable and simpler to understand. I urge you to do it, if not in this iteration, in the next one.

@aaclan-ebi and @amnonkhen I am aware of the Strategy Pattern, but in this case I am not sure how much would we win by applying it. The code would be even more complex just to remove not too many conditional statement. When you read about this pattern they also advice you not to use it every time as it could be too extensive to use in some cases.

amnonkhen commented:
> > Is there a test that checks that all the ENA files for a certain project are created correctly?
> 
> @amnonkhen What kind of ENA files? We are not dealing here with files. Are you talking about sample conversion? Sorry it is not really clear what would you like to know.

I am asking about a test that checks that a single project is sent to ENA properly. In other words that every piece of data that should move to ENA and every API call in ENA's submission framework is called.

ke4 commented:
> > > Is there a test that checks that all the ENA files for a certain project are created correctly?
> > 
> > 
> > @amnonkhen What kind of ENA files? We are not dealing here with files. Are you talking about sample conversion? Sorry it is not really clear what would you like to know.
> 
> I am asking about a test that checks that a single project is sent to ENA properly. In other words that every piece of data that should move to ENA and every API call in ENA's submission framework is called.

That is 2 things to test so they should be in 2 different test cases:
1. `test that checks that a single project is sent to ENA properly` -> this is tested in the various tests in the converter package
2. 'every API call in ENA's submission framework is called' -> This is tested in the submitter package

ke4 commented:
I merged it as @aaclan-ebi approved it earlier and her review just became stale as I added a dependency change and @amnonkhen approved it now. Tx
